### DISCUSSION TITLE
You can not use getInitialProps with getServerSideProps. Please remove getInitialProps.

### DISCUSSION THREAD

**Discussion by saeedPadyab** [vercel/next.js#11147] [Help]
Hi,
I have a HoC named withAuth which is gonna check the authentication for some pages.
first how can I write a hoc with getServerSideProps function? if can't, so what's the solution to have hoc which uses getInitialProps that wraps a page has got  getServerSideProps?

**Comment by travisluong** [comment]
This seems to have worked for me.

```
  Wrapper.getServerSideProps = async ctx => {
    const token = auth(ctx);

    const componentProps =
      WrappedComponent.getServerSideProps &&
      (await WrappedComponent.getServerSideProps(ctx));

    return { ...componentProps, token };
  };
```

**Reply by MrSaints** [reply]
I'm not sure how you managed this, but I get the following exception:

```
Error: page /<omitted> getServerSideProps can not be attached to a page's component and must be exported from the page. See more info here: https://err.sh/next.js/gssp-component-member
```

**Comment by MantaRafaela** [comment]
I have the same issue with @MrSaints 
Error: page /admin getServerSideProps can not be attached to a page's component and must be exported from the page. See more info here: https://err.sh/next.js/gssp-component-member

**Reply by todortotev** [reply]
@r-manta if you can provide a cloned repo it would be easiest to assist

**Comment by dsebastien** [comment]
Same issue here. Has anyone found a workaround?

**Reply by Sergioamjr** [reply]
I updated my project and tried to use the `getServerSideProps` method, however, I have a function that validates user access on a private page with an async call inside a `getInitialProps` method.

I solved this by creating a method that receives the context and a callback function, like a high order function, and I run it on all my private pages.

```js
// utils.js
const validateRoute = (context) => async (callback) => {
  const { id, pathname } = context.query;
  const paths = await getAllowedPaths(id);
  if (!paths.includes(pathname)) {
    return {
      redirect: "/home",
      permanente: false,
    };
  }
  return callback();
};
```

In my page.
```javascript
export const getServerSideProps = (context) => {
  return validateRoute(context)(async () => {
    // my custom fetch for each page.
    const { id } = context.query;
    const props = await getUserData(id);
    return {
      props,
    };
  });
};
```

I discovered that `getServerSideProps` method [is only available on pages componentes](https://nextjs.org/docs/basic-features/data-fetching#only-allowed-in-a-page-2).

If you need a HOC, I found some good ideas [here](https://github.com/vercel/next.js/discussions/10925).

**Reply by adefemi** [reply]
Pass a props to the hoc notifying it if a component implements a `getServerSideProps`, then you can ignore `getInitialProps` implementation for such components
